[IRCE] Relax profitability check

JanJecmen opened the pull request:
IRCE currently has two profitability checks:

1. min number of iterations (10 by default)
2. branch is highly biased (> 15/16)

However, it may still be profitable to eliminate range checks even if the branch isn't as biased. Consider, for example, a loop with 100 iterations, where IRCE currently eliminates all 100 range checks. The same range checks performed over a loop with 200 iterations aren't eliminated because the branch is 50-50.

This patch proposes to relax the profitability checks of IRCE. Namely, instead of the two checks currenly in place, consider IRCE profitable if the branch probability scaled by the expected number of iterations (i.e., the estimated number of eliminated checks) is over a threshold. This covers the minimum number of iterations check (there are at least as many iterations as eliminated range checks), and changes the bias check from a percent of iterations to at least a constant threshold of eliminated checks.

If the number of iterations can't be estimated, the check falls back to the current 15/16 likelihood check.

aleks-tmb commented on llvm/lib/Transforms/Scalar/InductiveRangeCheckElimination.cpp:
```diff
@@ -985,13 +981,25 @@ bool InductiveRangeCheckElimination::run(
 
   LLVMContext &Context = Preheader->getContext();
+
+  const char *FailureReason = nullptr;
```
Can we extract moving LoopStructure::parseLoopStructure into a separate NFC patch?

aleks-tmb commented:
> Consider, for example, a loop with 100 iterations, where IRCE currently eliminates all 100 range checks. The same range checks performed over a loop with 200 iterations aren't eliminated because the branch is 50-50.

Could you please add a test where range checks are not eliminated by the current IRCE, but are eliminated by your patch, to demonstrate the impact of the changes?

JanJecmen commented:
@aleks-tmb Hi, thanks for the review. I addressed your comments and pushed a new version. I tweaked the new test to better show the effect: IRCE previously wouldn't fire because the range check branch weights are 1:1, but now it eliminates 50 of the 100 checks. Setting the threshold to 51 then prevents it from applying. I also refactored some small details (used BranchProbability existing API for inversion, hoisted a loop invariant computation of the loop trip count estimate).

JanJecmen committed:
[IRCE] Relax profitability check

IRCE currently has two profitability checks:
  1. min number of iterations (10 by default)
  2. branch is highly biased (> 15/16)
However, it may still be profitable to eliminate range checks even
if the branch isn't as biased. Consider, for example, a loop with
100 iterations, where IRCE currently eliminates all 100 range checks.
The same range checks, if performed in a loop with 200 iterations,
are not eliminated because their branch is now only 1:1.

This patch proposes to relax the profitability checks of IRCE.
Namely, instead of the two checks currenly in place, consider IRCE
profitable if the branch probability scaled by the expected number
of iterations (i.e., the estimated number of eliminated checks)
is over a threshold. This covers the minimum number of iterations check
(there are at least as many iterations as eliminated range checks),
and changes the bias check from a percent of iterations to at least
a constant threshold of eliminated checks.
The effect is shown in the new test `profitability.ll`. The loop has
100 iterations (the backedge is taken 99:1). The range
check's branch weights are 1:1, so current IRCE wouldn't even
consider this a range check. However, with the new implementaion,
setting the minimum eliminated checks as high as 50, the
transformation is still applied.

If the number of iterations can't be estimated, the check falls back
to the current 15/16 likelihood check.

JanJecmen commented on llvm/lib/Transforms/Scalar/InductiveRangeCheckElimination.cpp:
```diff
@@ -940,35 +958,35 @@ PreservedAnalyses IRCEPass::run(Function &F, FunctionAnalysisManager &AM) {
 
-bool
-InductiveRangeCheckElimination::isProfitableToTransform(const Loop &L,
-                                                        LoopStructure &LS) {
```
Got it, done in #116384. Thanks!

aleks-tmb commented on llvm/lib/Transforms/Scalar/InductiveRangeCheckElimination.cpp:
```diff
@@ -989,17 +999,15 @@ bool InductiveRangeCheckElimination::run(
   }
 
-  if (!isProfitableToTransform(*L))
-    return false;
```
Currently, we have an early exit from the transformation if we determine it's not profitable. Does it make sense to keep this behavior? Is it a common path where we perform an early exit?

JanJecmen replied:
That is a good point, thanks. I added this early exit back

aleks-tmb commented:
LGTM
